src/zcash/JoinSplit.cpp

```diff
@@ -0,0 +1,345 @@
+        uint256& ephemeralKey,
+        const uint256& pubKeyHash,
+        uint256& randomSeed,
+        boost::array<uint256, NumInputs>& hmacs,
+        boost::array<uint256, NumInputs>& nullifiers,
+        boost::array<uint256, NumOutputs>& commitments,
+        uint64_t vpub_old,
+        uint64_t vpub_new,
+        const uint256& rt
+    ) {
+        if (!pk) {
+            throw std::runtime_error("JoinSplit proving key not loaded");
+        }
+
+        // Compute nullifiers of inputs
+        for (size_t i = 0; i < NumInputs; i++) {
+            nullifiers[i] = inputs[i].nullifier();
+        }
+
+        // Sample randomSeed
+        randomSeed = random_uint256();
+
+        // Compute h_sig
+        uint256 h_sig = this->h_sig(randomSeed, nullifiers, pubKeyHash);
+
+        // Sample phi
+        uint256 phi = random_uint256();
+
+        // Compute notes for outputs
+        for (size_t i = 0; i < NumOutputs; i++) {
+            // Sample r
+            uint256 r = random_uint256();
+
+            output_notes[i] = outputs[i].note(phi, r, i, h_sig);
+        }
+
+        // Compute the output commitments
+        for (size_t i = 0; i < NumOutputs; i++) {
+            commitments[i] = output_notes[i].cm();
+        }
+
+        // Encrypt the ciphertexts containing the note
+        // plaintexts to the recipients of the value.
+        {
+            ZCNoteEncryption encryptor(h_sig);
+
+            for (size_t i = 0; i < NumOutputs; i++) {
+                // TODO: expose memo in the public interface
+                boost::array<unsigned char, ZCASH_MEMO_SIZE> memo;
+                memo[0] = 0xF6; // invalid UTF8 as per spec
+
+                NotePlaintext pt(output_notes[i], memo);
+
+                ciphertexts[i] = pt.encrypt(encryptor, outputs[i].addr.pk_enc);
+            }
+
+            ephemeralKey = encryptor.get_epk();
+        }
+
+        // Authenticate h_sig with each of the input
+        // spending keys, producing hmacs which protect
+        // against malleability.
+        for (size_t i = 0; i < NumInputs; i++) {
+            hmacs[i] = PRF_pk(inputs[i].key, i, h_sig);
+        }
+
+        std::vector<FieldT> primary_input;
+        std::vector<FieldT> aux_input;
+
+        {
+            protoboard<FieldT> pb;
+            {
+                joinsplit_gadget<FieldT, NumInputs, NumOutputs> g(pb);
+                g.generate_r1cs_constraints();
+                g.generate_r1cs_witness(
+                    phi,
+                    rt,
+                    h_sig,
+                    inputs,
+                    output_notes,
+                    vpub_old,
+                    vpub_new
+                );
+            }
+
+            if (!pb.is_satisfied()) {
+                throw std::invalid_argument("Constraint system not satisfied by inputs");
+            }
+
+            primary_input = pb.primary_input();
+            aux_input = pb.auxiliary_input();
+        }
```

Comment:
These above lines are in structured blocks because the gadget isn't needed after the constraint system is realized, at least with the current design of the libsnark API, and the protoboard isn't needed once the input witnesses are realized. This should save us some memory; not sure how much.

Ultimately, I want to reuse the constraint system and input witnesses for the prover, because currently the proving key contains a copy of the constraint system. This is a limitation of the libsnark API that we should focus on closer to launch.